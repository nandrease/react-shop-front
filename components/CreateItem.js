import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';
import Form from './styles/Form';
import formatCurrency from '../lib/formatMoney';
import Error from './ErrorMessage';

const CREATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION(
		$title: String!
		$description: String!
		$price: Int!
		$image: String
		$largeImage: String
	) {
		createItem(title: $title, description: $description, price: $price, image: $image, largeImage: $largeImage) {
			id
		}
	}
`;

class CreateItem extends Component {
	state = {
		title: 'Cool shoes',
		description: 'Beautiful shoes',
		image: '',
		largeImage: '',
		price: 3400
	};
	handleChange = (e) => {
		const { name, type, value } = e.target;
		const val = type === 'number' && value ? parseFloat(value) : value;
		this.setState({ [name]: val });
	};

	uploadFile = async (e) => {
		const files = e.target.files;
		const data = new FormData();
		data.append('file', files[0]);
		data.append('upload_preset', 'shopreact');

		const res = await fetch('https://api.cloudinary.com/v1_1/doutr1bqe/image/upload', {
			method: 'POST',
			body: data
		});
		const file = await res.json();
		this.setState({
			image: file.secure_url,
			largeImage: file.eager[0].secure_url
		});
	};
	render() {
		return (
			<Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
				{(createItem, { loading, error }) => (
					<Form
						onSubmit={async (e) => {
							// stop form from submitting
							e.preventDefault();
							// call the mutation
							const res = await createItem();
							// change them to single item page
							Router.push({
								pathname: '/item',
								query: { id: res.data.createItem.id }
							});
						}}
					>
						<Error error={error} />
						<fieldset disabled={loading} aria-busy={loading}>
							<label htmlFor="file">
								Image
								<input
									type="file"
									id="file"
									name="file"
									placeholder="Upload an image"
									required
									onChange={this.uploadFile}
								/>
								{this.state.image && <img src={this.state.image} alt="Upload preview" width="200" />}
							</label>
							<label htmlFor="title">
								Title
								<input
									type="text"
									id="title"
									name="title"
									placeholder="Title"
									value={this.state.title}
									onChange={this.handleChange}
									required
								/>
							</label>
							<label htmlFor="price">
								Price
								<input
									type="number"
									id="price"
									name="price"
									placeholder="Price"
									value={this.state.price}
									onChange={this.handleChange}
									required
								/>
							</label>
							<label htmlFor="description">
								Description
								<textarea
									type="text"
									id="description"
									name="description"
									placeholder="Enter a description"
									value={this.state.description}
									onChange={this.handleChange}
									required
								/>
							</label>
							<button type="submit">Submit</button>
						</fieldset>
					</Form>
				)}
			</Mutation>
		);
	}
}

export default CreateItem;

export { CREATE_ITEM_MUTATION };
