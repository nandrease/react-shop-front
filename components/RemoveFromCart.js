import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { CURRENT_USER_QUERY } from './User';

const BigButton = styled.button`
	font-size: 3rem;
	background: none;
	border: 0;
	&:hover {
		color: ${(props) => props.theme.red};
		cursor: pointer;
	}
`;

const REMOVE_FROM_CART_MUTATION = gql`
	mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
		removeFromCart(id: $id) {
			id
		}
	}
`;

class RemoveFromCart extends Component {
	static propTypes = {
		id: PropTypes.string.isRequired
	};
	// This will get called as soon as we delete an item
	// and a response gets back from the server
	update = (cache, payload) => {
		// 1. first read the cache
		const data = cache.readQuery({ query: CURRENT_USER_QUERY });
		// 2. remove that item from the cart
		const cartItemId = payload.data.removeFromCart.id;
		data.me.cart = data.me.cart.filter((cartItem) => cartItem.id !== cartItemId);
		// 3. write it back to the cache
		cache.writeQuery({ query: CURRENT_USER_QUERY, data });
	};
	render() {
		const { id } = this.props;
		return (
			<Mutation
				mutation={REMOVE_FROM_CART_MUTATION}
				variables={{ id }}
				update={this.update}
				optimisticResponse={{ __typename: 'Mutation', removeFromCart: { __typename: 'CartItem', id } }}
			>
				{(removeFromCart, { loading, error }) => (
					<BigButton
						disabled={loading}
						onClick={() => {
							removeFromCart().catch((err) => alert(err.message));
						}}
						title="Delete Item"
					>
						&times;
					</BigButton>
				)}
			</Mutation>
		);
	}
}

export default RemoveFromCart;
export { REMOVE_FROM_CART_MUTATION };
