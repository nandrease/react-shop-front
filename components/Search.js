import React, { Component } from 'react';
import Downshift, {resetIdCounter} from 'downshift';
import Router from 'next/router';
import { ApolloConsumer } from 'react-apollo';
import gql from 'graphql-tag';
import debounce from 'lodash.debounce';
import { DropDown, DropDownItem, SearchStyles } from './styles/DropDown';

const SEARCH_ITEMS_QUERY = gql`
	query SEARCH_ITEMS_QUERY($searchTerm: String!) {
		items(where: { OR: [{ title_contains: $searchTerm }, { description_contains: $searchTerm }] }) {
			id
			image
			title
		}
	}
`;

const routeToItem = (item) => {
	Router.push({
		pathname: '/item',
		query: {
			id: item.id
		}
	});
};

class AutoComplete extends Component {
	state = {
		items: [],
		loading: false
	};
	onChange = debounce(async (e, client) => {
		this.setState({ loading: true });
		const res = await client.query({
			query: SEARCH_ITEMS_QUERY,
			variables: { searchTerm: e.target.value }
		});
		this.setState({
			items: res.data.items,
			loading: false
		});
	}, 350);
	render() {
		return (
			<SearchStyles>
				<Downshift onChange={routeToItem} itemToString={(item) => (item === null ? '' : item.title)}>
					{({ getInputProps, getItemProps, isOpen, inputValue, highlightedIndex }) => (
						<div>
							<ApolloConsumer>
								{(client) => (
									<input
										type="search"
										{...getInputProps({
											type: 'search',
											placeholder: 'Search for an Item',
											id: 'search',
											className: this.state.loading ? 'loading' : '',
											onChange: (e) => {
												e.persist();
												this.onChange(e, client);
											}
										})}
									/>
								)}
							</ApolloConsumer>
							{isOpen && (
								<DropDown>
									{this.state.items.map((item, index) => (
										<DropDownItem
											key={item.id}
											{...getItemProps({ item })}
											hightlighted={index === highlightedIndex}
										>
											<img width="50" src={item.image} alt={item.title} />
											<h3>{item.title}</h3>
										</DropDownItem>
									))}
									{!this.state.items.length &&
									!this.state.loading && <DropDownItem>Nothing found {inputValue}</DropDownItem>}
								</DropDown>
							)}
						</div>
					)}
				</Downshift>
			</SearchStyles>
		);
	}
}

export default AutoComplete;
