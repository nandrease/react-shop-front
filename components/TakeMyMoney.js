import React, { Component } from 'react';
import StripeCheckout from 'react-stripe-checkout';
import { Mutation } from 'react-apollo';
import Router from 'next/router';
import NProgress from 'nprogress';
import gql from 'graphql-tag';
import calcTotalPrice from '../lib/calcTotalPrice';
import Error from './ErrorMessage';
import User, { CURRENT_USER_QUERY } from './User';
import { create } from 'domain';

const CREATE_ORDER_MUTATION = gql`
	mutation CREATE_ORDER_MUTATION($token: String!) {
		createOrder(token: $token) {
			id
			charge
			total
			items {
				id
				title
			}
		}
	}
`;

function totalItems(cart) {
	return cart.reduce((tally, cartItem) => (tally = tally + cartItem.quantity), 0);
}

class TakeMyMoney extends Component {
	onToken = async (res, createOrder) => {
		NProgress.start();
		console.log('On token called');
		console.log(res.id);
		// manually call the mutation once we have the Stripe token
		const order = await createOrder({
			variables: {
				token: res.id
			}
		}).catch((err) => alert(err.message));
		console.log(order);
		Router.push({
			pathname: '/order',
			query: {
				id: order.data.createOrder.id
			}
		});
	};
	render() {
		return (
			<User>
				{({ data: { me }, loading }) => {
					if (loading) return null;
					return (
						<Mutation mutation={CREATE_ORDER_MUTATION} refetchQueries={[ { query: CURRENT_USER_QUERY } ]}>
							{(createOrder) => (
								<StripeCheckout
									amount={calcTotalPrice(me.cart)}
									name="Shop-react"
									description={`Order of ${totalItems(me.cart)}`}
									image={me.cart.length && me.cart[0].item && me.cart[0].item.image}
									stripeKey="pk_test_lhcc6UXxwByIR6M0gcfJPn3k"
									currency="EUR"
									email={me.email}
									token={(res) => this.onToken(res, createOrder)}
								>
									{this.props.children}
								</StripeCheckout>
							)}
						</Mutation>
					);
				}}
			</User>
		);
	}
}

export default TakeMyMoney;
export { CREATE_ORDER_MUTATION };
