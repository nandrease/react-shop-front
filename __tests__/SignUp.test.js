import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import { ApolloConsumer } from 'react-apollo';
import SignUp, { SIGNUP_MUTATION } from '../components/SignUp';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser } from '../lib/testUtils';

function type(wrapper, name, value) {
	wrapper.find(`input[name="${name}"]`).simulate('change', {
		target: { name, value }
	});
}

const me = fakeUser();
const mocks = [
	//signUp mock mutation
	{
		request: {
			query: SIGNUP_MUTATION,
			variables: {
				email: me.email,
				name: me.name,
				password: 'test'
			}
		},
		result: {
			data: {
				signUp: {
					__typename: 'User',
					id: 'abc123',
					email: me.email,
					name: me.name
				}
			}
		}
	},
	//current user query mock
	{
		request: { query: CURRENT_USER_QUERY },
		result: {
			data: { me }
		}
	}
];
describe('<SignUp />', () => {
	it('renders and matches snaphsot', () => {
		const wrapper = mount(
			<MockedProvider>
				<SignUp />
			</MockedProvider>
		);
		expect(toJSON(wrapper.find('form[data-test="form"]'))).toMatchSnapshot();
	});

	it('calls the mutation properly', async () => {
		let apolloClient;

		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<ApolloConsumer>
					{(client) => {
						apolloClient = client;
						return <SignUp />;
					}}
				</ApolloConsumer>
			</MockedProvider>
		);
		await wait();
		wrapper.update();
		type(wrapper, 'email', me.email);
		type(wrapper, 'name', me.name);
		type(wrapper, 'password', 'test');
		wrapper.update();
		wrapper.find('form[data-test="form"]').simulate('submit');
		await wait();

		//query the user ofut of the apollo client
		const user = await apolloClient.query({ query: CURRENT_USER_QUERY });
		expect(user.data.me).toMatchObject(me);
	});
});
