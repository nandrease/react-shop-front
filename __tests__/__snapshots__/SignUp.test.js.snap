// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> renders and matches snaphsot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  data-test="form"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      SignUp component
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="Enter your email"
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="name"
    >
      Name
      <input
        name="name"
        onChange={[Function]}
        placeholder="Enter your full name"
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        name="password"
        onChange={[Function]}
        placeholder="Enter a secure password"
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Sign Up!
    </button>
  </fieldset>
</form>
`;
